mne/epochs.py

```diff
@@ -3398,10 +3572,21 @@ def _concatenate_epochs(epochs_list, with_data=True, add_offset=True, *,
     if not isinstance(epochs_list, (list, tuple)):
         raise TypeError('epochs_list must be a list or tuple, got %s'
                         % (type(epochs_list),))
+
+    # to make warning messages only occur once during concatenation
+    warned = False
+
     for ei, epochs in enumerate(epochs_list):
         if not isinstance(epochs, BaseEpochs):
             raise TypeError('epochs_list[%d] must be an instance of Epochs, '
                             'got %s' % (ei, type(epochs)))
+
+        if (getattr(epochs, 'annotations', None) is not None and
+                len(getattr(epochs, 'annotations', None)) != 0) and not warned:
```

Comment:
`and` is a short-circuiting operation, so by this point in the conditional we know the first part has ben satisfied
```suggestion
        if (getattr(epochs, 'annotations', None) is not None and
                len(epochs.annotations) > 0 and
                not warned):
```